### PR TITLE
feat(memory): add listThreads method with flexible filtering

### LINKED ISSUE #4333: [FEATURE] Query messages based on metadata
### Problem Statement

As our agent system grows in complexity, we've found it useful to start using `metadata` on threads more and more, e.g. to categorize types of threads.

It would be quite useful to have a way to query threads based on the metadata. If not exposed through Mastra, we would need to create our own tables to keep track of this. Given that Mastra provides the functionality to manage metadata, it would make sense to offer this functionality as well.

Potentially related to this issue around creating pagination on threads (and messages): https://github.com/mastra-ai/mastra/issues/3661

### Proposed Solution

I'm not sure, but something like:

```javascript
mastraClient.getMemoryThreads({resourceId, agentId: agentId, {
  metadata: {
    contains: "categoryA"
  }
}});
}});
```

I imagine this could easily be over-complicated, so having some simple but powerful interface of querying would be great.

### Component

Memory

### Alternatives Considered

_No response_

### Example Use Case

Metadata on threads are useful for a number of things including categorizing threads (e.g. some threads might be focused on specific areas).

Having an easy way to pull out threads for a certain category would be useful.

### Additional Context

_No response_

### Verification

- [x] I have searched the existing issues to make sure this is not a duplicate
- [x] I have provided sufficient context for the team to understand the request

### PR DESCRIPTION
**Pull request by roaminro**
## Summary

Implements the new `listThreads` method across all storage adapters, based on [PR #10662](https://github.com/mastra-ai/mastra/pull/10662).

 closes #4333

### Features
- **Flexible filtering**: Optional `resourceId` and `metadata` filters (both optional, unlike `listThreadsByResourceId`)
- **Metadata filtering**: AND logic - all specified key-value pairs must match
- **Pagination & sorting**: Same options as `listThreadsByResourceId`

### Security Improvements 🔒
- **SQL Injection Prevention**: Added `validateMetadataKeys()` to prevent SQL injection via malicious metadata keys
  - Keys must match pattern: `^[a-zA-Z_][a-zA-Z0-9_]*$`
  - Maximum key length of 128 characters enforced
- **Integer Overflow Protection**: Added `validatePagination()` to prevent overflow attacks on page parameters

### Storage Adapters Updated (11 total)
| Category | Stores |
|----------|--------|
| SQL-based | PostgreSQL, LibSQL, MSSQL, Cloudflare D1, ClickHouse |
| NoSQL/KV | MongoDB, DynamoDB, Upstash, Cloudflare KV, Convex, Lance |

### Usage Example

```typescript
// Filter by resourceId only
await memory.listThreads({ filter: { resourceId: 'user-123' } });

// Filter by metadata only
await memory.listThreads({ filter: { metadata: { category: 'support' } } });

// Filter by both (AND logic)
await memory.listThreads({
  filter: {
    resourceId: 'user-123',
    metadata: { priority: 'high', status: 'open' }
  },
  orderBy: { field: 'updatedAt', direction: 'DESC' },
  page: 0,
  perPage: 20
});
```

## Test Plan
- [x] Type checking passes across all packages
- [x] Comprehensive test coverage for filtering, pagination, and sorting
- [x] Run full test suite
- [x] Verify SQL injection prevention works as expected
- [x] Test with edge cases (empty metadata, special characters rejected)

## Summary by CodeRabbit

* **New Features**
  * Unified listThreads API: optional resourceId and metadata filters (AND semantics), pagination with fetch-all option, and ordering support.

* **Improvements**
  * Stricter pagination validation, clearer page/perPage/total/hasMore semantics, and metadata key/value validation with richer error/context.
  * Added deep JSON equality utility for comparisons.

* **Documentation**
  * Updated guides, references, examples, and redirects for the new listing API.

* **Tests & Migration**
  * Expanded test coverage and a codemod to assist migration.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by roaminro** [comment]
**Re: JSON.parse in metadata schema (memory.ts:194)**
✅ Fixed in commit 458b957543.

Wrapped `JSON.parse` in try/catch for both `listThreadsQuerySchema` and `listThreadsNetworkQuerySchema`:
- Returns `undefined` on parse error instead of throwing
- Zod validation produces proper 400 error instead of 500
- Matches pattern used by other schemas (storageOrderBySchema, messageOrderBySchema, etc.)

**Comment by roaminro** [comment]
**Re: Filtering tests nested in sorting block (threads.ts:800)**
📝 Cancelled/Not applicable.

After investigation, these filtering tests are not nested inside conditional `describeSorting` blocks. The test structure is correct and filtering tests run for all storage adapters. No changes needed.

**Comment by roaminro** [comment]
**Status Update on Remaining Items:**

**🟡 Minor (Acknowledged, can be follow-up):**
- MockStorage types (`message-history.test.ts`) - Works correctly but could use proper types
- Negative test cases (`memory.test.ts`) - Schemas handle errors correctly, but explicit tests would improve coverage  
- Filtering tests organization - Verified these are NOT nested incorrectly; tests run for all adapters

**🔴 DynamoDB Store (Deferred to separate PR):**
- ElectroDB API usage and TypeScript strict mode errors
- Pagination validation for `perPage: false`
- Logger sanitization (avoid logging PII)
- Metadata filtering for stringified JSON

These are store-specific optimizations requiring DynamoDB-specific testing. They don't block the core listThreads functionality which works correctly across all other storage adapters.

File: docs/src/content/en/guides/migrations/upgrade-to-v1/storage.mdx [hunk 1]
```diff
@@ -150,28 +150,57 @@ To migrate, call `getStore()` with the domain name, then call methods on the ret
 + const scores = await scoresStore?.listScoresByScorerId({ scorerId: 'helpfulness' });
 ```
 
-### `getThreadsByResourceId` to `listThreadsByResourceId`
+### `getThreadsByResourceId` to `listThreads`
```
Comment (LekoArts) [inline docs/src/content/en/guides/migrations/upgrade-to-v1/storage.mdx:194]: This migration doesn't really explain how the user has to handle the pagination now. Previously they were able to get all threads, now the examples below show pagination. That's a changed behavior that needs to be addressed

File: docs/src/content/en/guides/migrations/upgrade-to-v1/storage.mdx [hunk 2]
```diff
@@ -150,28 +150,57 @@ To migrate, call `getStore()` with the domain name, then call methods on the ret
...
+});
 ```
 
 :::tip[Codemod]
 
 You can use Mastra's codemod CLI to update your code automatically:
 
 ```shell
-npx @mastra/codemod@beta v1/storage-get-threads-by-resource .
+npx @mastra/codemod@latest list-threads-by-resource-to-list-threads .
```
Comment (LekoArts) [inline docs/src/content/en/guides/migrations/upgrade-to-v1/storage.mdx]:
```suggestion
npx @mastra/codemod@beta v1/list-threads-by-resource-to-list-threads .
```

File: docs/src/content/en/guides/migrations/upgrade-to-v1/storage.mdx [hunk 3]
```diff
@@ -150,28 +150,57 @@ To migrate, call `getStore()` with the domain name, then call methods on the ret
 + const scores = await scoresStore?.listScoresByScorerId({ scorerId: 'helpfulness' });
 ```
 
-### `getThreadsByResourceId` to `listThreadsByResourceId`
+### `getThreadsByResourceId` to `listThreads`
 
-The `getThreadsByResourceId()` method has been renamed to `listThreadsByResourceId()`. This change aligns with the convention that `list*` methods return collections.
+The `getThreadsByResourceId()` method has been replaced with `listThreads()`, which supports optional filtering by `resourceId` and/or `metadata`.
```
Comment (LekoArts) [inline docs/src/content/en/guides/migrations/upgrade-to-v1/storage.mdx]:
```suggestion
The `getThreadsByResourceId()` method has been replaced with `listThreads()`. `listThreads()` supports filtering by `resourceId` or `metadata` or a combination of both.
```

File: docs/src/content/en/reference/memory/listThreads.mdx [hunk 4]
```diff
@@ -0,0 +1,196 @@
+---
+title: "Reference: Memory.listThreads() | Memory"
+description: "Documentation for the `Memory.listThreads()` method in Mastra, which retrieves threads with optional filtering by resourceId and/or metadata."
+packages:
+  - "@mastra/memory"
+---
+
+# Memory.listThreads()
+
+The `.listThreads()` method supports filtering threads by `resourceId` and/or `metadata`, and replaces the deprecated `listThreadsByResourceId()` method.
```
Comment (LekoArts) [inline docs/src/content/en/reference/memory/listThreads.mdx]:
```suggestion
The `listThreads()` method returns all thread and supports filtering by `resourceId` or `metadata` or a combination of both.
```

File: docs/src/content/en/reference/memory/listThreads.mdx [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+## Usage Examples
```
Comment (LekoArts) [inline docs/src/content/en/reference/memory/listThreads.mdx:12]: Same as above, all these examples show pagination

File: docs/src/content/en/reference/memory/listThreads.mdx [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+  page: 0,
+  perPage: 10,
+});
+```
+
+A codemod is available to automate this migration:
+
+```bash
+npx @mastra/codemod@latest list-threads-by-resource-to-list-threads
+```
```
Comment (LekoArts) [inline docs/src/content/en/reference/memory/listThreads.mdx]: That needs to be removed, the migration guide exists for this

File: packages/codemod/src/codemods/v1/storage-list-threads-by-resource-to-list-threads.ts
Comment (LekoArts) [inline packages/codemod/src/codemods/v1/storage-list-threads-by-resource-to-list-threads.ts:1]: Did you use `pnpm scaffold` to create this codemod?

Because you're missing an entry in packages/codemod/src/lib/bundle.ts

**Review by LekoArts — CHANGES_REQUESTED** [review]

File: docs/src/content/en/guides/migrations/upgrade-to-v1/storage.mdx [hunk 7]
```diff
@@ -191,28 +191,69 @@ To migrate, call `getStore()` with the domain name, then call methods on the ret
...
 ```
 
-### `getThreadsByResourceId` to `listThreadsByResourceId`
+### `getThreadsByResourceId` to `listThreads`
 
-The `getThreadsByResourceId()` method has been renamed to `listThreadsByResourceId()`. This change aligns with the convention that `list*` methods return collections.
+The `getThreadsByResourceId()` method has been replaced with `listThreads()`. The new method adds pagination support and filtering by `resourceId`, `metadata`, or both.
 
-To migrate, use the memory store and rename the method call with pagination parameters.
+**Important:** The old `getThreadsByResourceId()` returned all matching threads without pagination. The new `listThreads()` requires pagination parameters. To preserve the old behavior of fetching all threads, use `perPage: false`.
```
Comment (LekoArts) [inline docs/src/content/en/guides/migrations/upgrade-to-v1/storage.mdx]:
```suggestion
:::important

The old `getThreadsByResourceId()` returned all matching threads without pagination. The new `listThreads()` requires pagination parameters. To preserve the old behavior of fetching all threads, use `perPage: false`.

:::
```

File: docs/src/content/en/guides/migrations/upgrade-to-v1/storage.mdx [hunk 8]
```diff
@@ -191,28 +191,69 @@ To migrate, call `getStore()` with the domain name, then call methods on the ret
...
+
+To migrate, use the memory store and the new `listThreads()` method with pagination and an optional filter object.
 
 ```diff
```
Comment (LekoArts) [inline docs/src/content/en/guides/migrations/upgrade-to-v1/storage.mdx:206]: Much better 👍

File: docs/src/content/en/reference/memory/listThreads.mdx [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+```typescript
+const result = await memory.listThreads({
+  page: 0,
+  perPage: 10,
+});
+```
+
+### Fetch all threads without pagination
+
+Use `perPage: false` to retrieve all matching threads at once.
```
Comment (LekoArts) [inline docs/src/content/en/reference/memory/listThreads.mdx:25]:
```suggestion
Use `perPage: false` to retrieve all matching threads at once.

:::warning

Generally speaking it's recommended to use pagination, especially for large datasets. Use this option cautiously.

:::
```

File: docs/src/content/en/reference/memory/listThreads.mdx [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+```typescript
+const result = await memory.listThreads({
+  filter: { metadata: { category: "support", priority: "high" } },
+  page: 0,
+  perPage: 10,
+});
+```
+
+### Combined filter (resourceId AND metadata)
```
Comment (LekoArts) [inline docs/src/content/en/reference/memory/listThreads.mdx]:
```suggestion
### Combined filter (resourceId & metadata)
```

**Review by LekoArts — CHANGES_REQUESTED** [review]
Mostly checked the code tests, they look good to me. So only some small docs changes needed

**Review by LekoArts — APPROVED** [review]
Another pair of eyes 👁️ would be good since this is a bigger change
